server/src/main/scala/org/ensime/server/DocsHandler.scala

```diff
@@ -0,0 +1,98 @@
+import io.netty.handler.codec.http.HttpMethod.GET
+import io.netty.handler.codec.http.HttpResponseStatus
+import io.netty.handler.codec.http.HttpResponseStatus.{ BAD_REQUEST, FORBIDDEN, NOT_FOUND, OK }
+import io.netty.handler.codec.http.HttpVersion.HTTP_1_1
+import io.netty.handler.codec.http.{ DefaultFullHttpResponse, FullHttpRequest, HttpUtil }
+import io.netty.util.CharsetUtil
+
+import org.ensime.core.DocJarReading
+
+class DocsHandler(docs: DocJarReading) extends SimpleChannelInboundHandler[FullHttpRequest] {
+
+  val docRegex = """/docs/([^/]+\.jar)/(.+)""".r
+
+  override def channelRead0(ctx: ChannelHandlerContext, req: FullHttpRequest): Unit = {
+
+    // Handle a bad request.
+    if (!req.decoderResult().isSuccess())
+      sendHttpResponse(ctx, req, Left(BAD_REQUEST))
+
+    // Allow only GET methods.
+    else if (req.method() != GET)
+      sendHttpResponse(ctx, req, Left(FORBIDDEN))
+
+    // Send the index page listing the doc jars
+    else if ("/docs".equals(req.uri())) {
+      val content: ByteBuf = docJarsPage
+      sendHttpResponse(ctx, req, Right((Some("text/html"), content)))
+
+      // Send the content for the requested filename and entry pair
+    } else if (req.uri().startsWith("/docs/")) {
+      val docRegex(filename, entry) = req.uri()
+      docs.docJarContent(filename, entry)
+        .fold(sendHttpResponse(ctx, req, Left(NOT_FOUND))) {
+          jarContent =>
+            val mediaOpt = Option(Files.probeContentType(Paths.get(entry)))
+            val content = Unpooled.copiedBuffer(jarContent.toArray)
+            sendHttpResponse(ctx, req, Right((mediaOpt, content)))
+        }
+
+      // Can't handle this request
+    } else sendHttpResponse(ctx, req, Left(NOT_FOUND))
+  }
+
+  private def sendHttpResponse(
+    ctx: ChannelHandlerContext, req: FullHttpRequest,
+    statusOrContent: Either[HttpResponseStatus, (Option[String], ByteBuf)]
+  ): Unit = {
+
+    val res = statusOrContent match {
+      case Left(status) =>
+        // Generate an error page if response status code is not OK (200).
+        val res = new DefaultFullHttpResponse(HTTP_1_1, status)
+        val buf: ByteBuf = Unpooled.copiedBuffer(status.toString(), CharsetUtil.UTF_8)
+        res.content().writeBytes(buf)
+        buf.release()
+        res
+      case Right((mimeTypeOpt, content)) =>
+        val res = new DefaultFullHttpResponse(HTTP_1_1, OK, content)
+        mimeTypeOpt.foreach {
+          mimeType => res.headers().set(CONTENT_TYPE, s"$mimeType; charset=UTF-8")
+        }
+        res
+    }
+    HttpUtil.setContentLength(res, res.content().readableBytes())
+
+    // Send the response and close the connection if necessary.
+    val f: ChannelFuture = ctx.channel().writeAndFlush(res)
+    if (!HttpUtil.isKeepAlive(req) || res.status().code() != OK.code())
+      f.addListener(ChannelFutureListener.CLOSE)
+  }
+
+  private val docJarsPage: ByteBuf = {
+    val content =
+      <html>
```

Comment:
There was a project to do xml interpolation but I'm not sure it went anywhere. There is also Li Haoyi's tags library.